### ISSUE TITLE
Add option for signed S3 payload upload for S3 backends

### ISSUE THREAD

**Issue by bitbier** [opentofu/opentofu#1943]
### OpenTofu Version

```shell
OpenTofu v1.8.1
```

### Use Cases

We currently have buckets that enforce specific restrictions on the objects being uploaded. Specifically, we have buckets with the following policies:

```
        {
            "Effect": "Deny",
            "Principal": "*",
            "Action": "s3:PutObject",
            "Resource": "arn:aws:s3:::${BUCKET_NAME}/*",
            "Condition": {
                "StringEquals": {
                    "s3:x-amz-content-sha256": "UNSIGNED-PAYLOAD"
                }
            }
        },
```

The policy is to prevent users from uploading unsigned objects to our S3 bucket.

https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
https://docs.aws.amazon.com/AmazonS3/latest/API/bucket-policy-s3-sigv4-conditions.html

However, because of this policy, we cannot use backend S3 without reducing our security posture. This is because the current backend implementation for S3 does not sign the payload. Currently, there is no workaround for this as it looks like additional code is required to implement this signing.

We found this out by investigating the debug log when tofu tries to modify a state file (for our example, we were just importing an existing object).

When we did this, we saw:

```
2024-08-26T11:42:11.141-0700 [DEBUG] backend-s3: ... http.request.header.x_amz_content_sha256=UNSIGNED-PAYLOAD ...
```

Anyways, it looks like aws-sdk-go-v2 does not make this easy. The default implementation seems to aways set this header to `UNSIGNED-PAYLOAD` for whatever reason: https://github.com/aws/aws-sdk-go-v2/issues/1689

To make this work with signed payloads, looks like you might need to add some middleware for this to work: https://github.com/aws/aws-sdk-go-v2/issues/1566#issuecomment-1278238025

### Attempted Solutions

A workaround is to remove this policy on the bucket policy and enforce other security restrictions on the requests being made to the bucket. However, again, this may reduce your security posture if other counter measures are not used.

**Comment by skyzyx** [comment]
@bitbier: When you say "signing", in the context of https://github.com/aws/aws-sdk-go-v2/issues/1689, it sounds like you're just talking about calculating the SHA digest. Am I misunderstanding?

While I understand the value of checksumming, I'd also wonder if in-flight and at-rest SSE would give you a better guarantee. And that seems easier to do.

**Comment by bitbier** [comment]
Yes, correct. It is generating the checksum of the file that you are uploading to S3 so that S3 can verify the integrity of the file. The reason why we have the bucket policy above is to prevent users from uploading data via an presigned URL or via authentication headers. This is recommended by AWS for added security. Our security posture is to do all of the above. We also have SSE and enforce encryption in transit as well.
